docs/production/modifying-zulip.md

```diff
@@ -0,0 +1,206 @@
+
+If you are using [docker-zulip][], there are two things that are
+different from the above:
+
+* Because of how container images work, editing files directly is an
+  even worse idea, because Docker is designed for working with
+  container images and may lose your changes.
+* Instead of running `upgrade-zulip-from-git`, you will need to use
+  the [docker upgrade workflow][docker-zulip-upgrade] to build a
+  container image based on your modified version of Zulip.
+
+[docker-zulip]: https://github.com/zulip/docker-zulip
+[docker-zulip-upgrade]: https://github.com/zulip/docker-zulip#upgrading-from-a-git-repository
+
+## Applying changes from master
+
+If you are experiencing an issue that has already been fixed by the
+Zulip development community, and you'd like to get the fix now, you
+have a few options.  There are two possible ways you might get those
+fixes on your local Zulip server without waiting for an official release.
+
+### Applying a small change
+
+Many bugs have small/simple fixes.  In this case, you can use the Git
+workflow [described above](#making-changes), using:
+
+```
+git fetch upstream
+git cherry-pick abcd1234
+```
+
+instead of "making changes locally" (where `abcd1234` is the commit ID
+of the change you'd like).
+
+In general, we can't provide community support for issues caused by
+cherry-picking changes in this way.  There is a major exception to
+this rule.
+
+Because Zulip can be deployed in a wide variety of configurations and
+environments, we often merge fixes to bugs that we can reproduce in an
+automated test but have not directly reproduced in a production
+system.  In these cases, we'll ask the user(s) who reported the bug to
+apply the patch to their production system to verify the fix works for
+them.
+
+Generally, we'll only request this if we expect the fix in question to
+apply cleanly to the latest release without introducing regressions,
+and you can expect the Zulip community to be responsive in debugging
+any problems any caused by the patch.
+
+### Upgrading to master
+
+It's unsafe to backport arbitrary patches from master to an older
+version.  Common issues include:
+
+* Changes containing database migrations (new files under
+  `zerver/migrations/`), which includes most new major features.  We
+  don't support applying database migrations out of order.
+* Changes that are stacked on top of other changes to the same system
+  (generally this will be clear because).
+* Essentially any patch with hundreds of lines of changes.
+
+While it's possible to backport these sorts of changes, you're
+unlikely to succeed without help from the core team via a support
+contract.
+
+If you need an unreleased feature, the right path is usually to Zulip
+master using `[upgrade-zulip-from-git][].  Before upgrading to master,
+make sure you understand:
+
+* Master is our main development branch under very active development
+  and has dozens of new changes added most days.  Master can have
```

Comment:
Should lowercase `master`, or perhaps avoid the awkward sentence casing issue with “the `master` branch”.